### PR TITLE
Add PPUCD enumerated property parsing

### PR DESCRIPTION
**Pull request by echeran**
Based against the branch for PR #439, which is in turn based on PR #242 . The last 4 commits in this PR are added on top of #439 .

Also includes a fix for binary property parsing that supports the "-" prefix notation where "-WSpace" is a value that sets "WSpace" to false in PPUCD (important due to the cascading/overriding nature of PPUCD info).

### CODE REVIEW

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 1]
```diff
@@ -0,0 +1,1107 @@
...
+}
+
+fn is_enum_val_line(line: &str) -> bool {
+    line.starts_with("value;")
+}
+
+/// For a property definition or enumerated property value line, update the aliases map.
+fn update_aliases<'s>(prop_aliases: &mut HashMap<&'s str, HashSet<&'s str>>, line: &'s str) {
+    let mut line_parts = split_line(&line);
+    line_parts.drain(0..2);
```
Comment (echeran) [inline components/provider_ppucd/src/parse_ppucd.rs:61]: Use slice syntax for `.drain()`

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 2]
```diff
@@ -0,0 +1,1107 @@
...
+    // property;Binary;;blank
+    // property;Binary;;graph
+    // property;Binary;;print
+    // property;Binary;;xdigit
+    if line_parts[0].is_empty() {
+        line_parts.drain(0..1);
+    }
+
+    let canonical_name = line_parts[0];
+    let all_names: Vec<&'s str> = line_parts.iter().copied().collect();
```
Comment (echeran) [inline components/provider_ppucd/src/parse_ppucd.rs:75]: combine the iter -> Vec -> HashSet into one line

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 3]
```diff
@@ -0,0 +1,1107 @@
...
+            }
+        }
+    }
+
+    let mut result: HashMap<&str, UnicodeSet> = HashMap::new();
+
+    for (canonical_prop_name, prop_val_builder_map) in m {
+        for (canonical_val_name, uniset_builder) in prop_val_builder_map {
+            let enum_val_uniset_name: String =
+                format!("{}={}", canonical_prop_name, canonical_val_name);
```
Comment (echeran) [inline components/provider_ppucd/src/parse_ppucd.rs:355]: 3 options provided by Shane:

1. Use a substring from ppucd.txt itself in the `value;<propname>;<propval>'...` line in the header section
2. Use a TinyString
3. Use a Cow (and construct it with `Cow::Owned` b/c the input is a `String` resulting from `format!`)

Tentative approach:
Go with option 3 (use a Cow), and leave TinyString alternative for #454 optimization issue work. Also, make replacing `format!` a task item to consider in #454.

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 4]
```diff
@@ -48,33 +52,102 @@ fn is_property_line(line: &str) -> bool {
...
+    prop_vals: HashMap<&'s str, &'s str>,
+) -> HashMap<&'s str, &'s str> {
+    let mut prop_names: HashSet<&'s str> = prop_vals.keys().copied().collect();
+    // If we see "-Gr_Base", then remove both "Gr_Base" and "-Gr_Base".
+    for prop_name in prop_vals.keys() {
+        if let Some(unadorned_prop_name) = prop_name.strip_prefix('-') {
+            let orig_prop_name: &str = unadorned_prop_name;
+            prop_names.remove(&orig_prop_name);
+            prop_names.remove(prop_name);
+        }
```
Comment (nordzilla) [inline components/provider_ppucd/src/parse_ppucd.rs:142]: Drive-by: 

Having both `unadored_prop_name` and `orig_prop_name` seems unnecessary. 

Using just one of them should work fine:
```rust
    for prop_name in prop_vals.keys() {
        if let Some(orig_prop_name) = prop_name.strip_prefix('-') {
            prop_names.remove(orig_prop_name);
            prop_names.remove(prop_name);
        }
    }
```

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 5]
```diff
@@ -48,33 +52,102 @@ fn is_property_line(line: &str) -> bool {
...
+    }
+    prop_vals
+        .clone()
+        .into_iter()
+        .filter(|(k, _)| prop_names.contains(k))
+        .collect()
```
Comment (nordzilla) [inline components/provider_ppucd/src/parse_ppucd.rs:148]: Drive-by: 

It looks like you've got some unnecessary allocations here.

The first is the `.clone()`. Since the map is being moved in and then moved out, cloning the map before calling `into_iter()` will construct a new map, then move the new map into the iterator, when we could just move the existing map into the iterator. 

In general, re-collecting into a new map by moving into an iterator may also cause unnecessary allocations (unless the compiler is smart enough to see what's going on an optimize it to use the existing map anyway, which would be awesome and impressive).

I recommend using the `retain()` method, which will just filter the keys within the existing hash map.

```rust
fn apply_exclude_vals_for_binary_props<'s>(
    mut prop_vals: HashMap<&'s str, &'s str>,
) -> HashMap<&'s str, &'s str> {
    // ...
    // HashSet stuff here
    // ...
    prop_vals.retain(|prop_name, _| {
        prop_names.contains(prop_name)
    });
    prop_vals
}
```

Note, you'll have to make the `prop_vals` argument `mut` to use `retain()` on it.

**Review by nordzilla — COMMENTED** [review]
I saw this pop up in my email and glanced at it (though I didn't read through the entire PR). One of the functions caught my eye and I left a couple comments.

**Review by zbraniecki — APPROVED** [review]

File: components/provider_ppucd/src/parse_ppucd.rs (hunk 4 above)
Comment (echeran) [reply components/provider_ppucd/src/parse_ppucd.rs:138]: Done.

File: components/provider_ppucd/src/parse_ppucd.rs (hunk 5 above)
Comment (echeran) [reply components/provider_ppucd/src/parse_ppucd.rs:144]: I tried this change out, but then I got the "[Cannot borrow `x` as mutable more than once at a time](https://stackoverflow.com/questions/31281155/cannot-borrow-x-as-mutable-more-than-once-at-a-time)" type of error at the call sites for this function.  The code there in `get_code_point_prop_vals` looks like
```
...
    let mut prop_vals: HashMap<&'s str, &'s str> = HashMap::new();
    prop_vals.clone_from(defaults);
    for (range, block_prop_vals) in blocks {
        if range.contains_u32(code_point) {
            prop_vals.extend(block_prop_vals);
            prop_vals = apply_exclude_vals_for_binary_props(prop_vals);
        }
    }
```
and the compiler error complains about the 2 references to `prop_vals` being in the same lexical scope.

Is there a way to work around this? Also, I should have mentioned this issue in the followup issue #454 for improving performance, so if you have ideas, feel free to add them there.

File: components/uniset/src/provider.rs [hunk 6]
```diff
@@ -81,7 +81,7 @@ pub mod key {
     pub const XIDS_V1: ResourceKey = resource_key!(uniset, "XIDS", 1);
 }
 
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Hash, Eq)]
```
Comment (sffc) [inline components/uniset/src/provider.rs:84]: It looks like you only need `Hash` and `Eq` for testing, right?  You might as well feature-gate these derives then.  When you update your branch, you should see an example of how to do that when I added Serde as optional.

**Review by sffc — COMMENTED** [review]
Do you also intend to include the updates to uniset::props in this PR?

File: components/provider_ppucd/src/parse_ppucd.rs (hunk 5 above)
Comment (echeran) [reply components/provider_ppucd/src/parse_ppucd.rs]: I take back what I said -- I was getting confused by the compiler's suggestions and to which side of the colon the `mut` belongs not knowing the implications of the difference. Your suggested change did indeed work.

Btw, I made an extra modification on top of that -- I changed the signature from taking `mut prop_vals: HashMap<&'s str, &'s str>` with a return type of `HashMap<&'s str, &'s str>` to having no return value (since it's now just mutating the map) and taking `prop_vals: &mut HashMap<&'s str, &'s str>`.

**Comment by echeran** [comment]
> Do you also intend to include the updates to uniset::props in this PR?

Ah, thanks for reminding me. I just added the getter fns and the `ResourceKey`s that they require.

However, I am running into a problem that some of the resource keys' subcategory name string representations are too big for `TinyStr16`, ex:
```
error: proc macro panicked
   --> components/uniset/src/provider.rs:262:75
    |
262 |     pub const INDIC_POSITIONAL_CATEGORY_BOTTOM_AND_LEFT_V1: ResourceKey = resource_key!(uniset, "InPC=Bottom_And_Left", 1);
    |                                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
```

(That's usually not the case, ex: `WORD_BREAK_MIDLETTER_V1` has a string representation `"WB=ML"`.)

I've created the strings to fit the pattern `{prop_name_short_name}={prop_val_short_name}`. But since I think these strings are more or less an internal implementation that users need not know about to use the provider, should I manually truncate these strings to fit in 16 chars? Another option is to make the `subcateogry` field of ResourceKey be a `TinyStrAuto`.

File: components/provider_ppucd/src/parse_ppucd.rs (hunk 5 above)
Comment (nordzilla) [reply components/provider_ppucd/src/parse_ppucd.rs]: Sorry @echeran I was out the past two days. 

Glad you got this working. 

Re: your extra modification,

Totally! I think it's probably more typical to just take a `&mut HashMap`, but sometimes the move-in, move-out semantics are more desirable. I didn't look  further into the context of its usage to form an opinion about it.

**Comment by echeran** [comment]
I truncated the sub-category strings used internally when constructing the fixed `ResourceKey`s for enum properties.

In the meantime, I also had a suggestion to change the enum getter fn API to take an enum (ex: `get_numeric_type_digit`, `get_numeric_type_decimal`, ... => `get_numeric_type_property_value_set(NumericType.DIGIT)`, `get_numeric_type_property_value_set(NumericType.DECIMAL)`.  However, I understand that it would create some code/data dependencies that make dead code elimination more difficult. So I'll hold off for now. Also, we can always add that API formulation to provide for developer user convenience at a later point when needed.

File: components/uniset/src/provider.rs (hunk 6 above)
Comment (sffc) [reply components/uniset/src/provider.rs:901]: Ping

File: components/uniset/src/provider.rs [hunk 7]
```diff
@@ -14,74 +14,891 @@ pub mod key {
...
+    pub const INDIC_POSITIONAL_CATEGORY_NA_V1: ResourceKey = resource_key!(uniset, "InPC=NA", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_OVERSTRUCK_V1: ResourceKey =
+        resource_key!(uniset, "InPC=Overstruck", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_RIGHT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=Right", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_TOP_V1: ResourceKey = resource_key!(uniset, "InPC=Top", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_TOP_AND_BOTTOM_V1: ResourceKey =
+        resource_key!(uniset, "InPC=TopBottom", 1);
+    pub const INDIC_POSITIONAL_CATEGORY_TOP_AND_BOTTOM_AND_LEFT_V1: ResourceKey =
+        resource_key!(uniset, "InPC=TopBotLeft", 1);
```
Comment (sffc) [inline components/uniset/src/provider.rs:322]: This looks like one of the keys you shortened to fit in 16 characters.  Can you add a test to assert that PpucdDataProvider returns valid data when you ask for a key like this?

Comment (echeran) [reply components/uniset/src/provider.rs:322]: I added the test, but it reveals that the manual truncation of the subcategory string to fit in 16 characters actually causes an error. The `ResourceKey`'s subcategory string is used to select the correct `UnicodeProperty` from the PPUCD provider via `UnicodeProperty`'s `name` field. The `name` field is computed automatically by the parser as `<propname>=<propval>`.

The 16 character restriction is a problem in uniquely defining the specific UnicodeProperty (aka UnicodeSet) for a specific enum property name=value combination.

I guess the options are
1. revisit the TinyStr16 type for subcategory that causes 16 char limit
2. create a map that stores the mapping of  the new manual truncation string to the original `<propname>=<propval>` string, and use that map when doing a lookup of the correct `UnicodeProperty` from the data provider when given a ResourceKey

Option 2 looks preferable.

Comment (zbraniecki) [reply components/uniset/src/provider.rs:322]: In particular, we can make option 2 efficient by only storing maps for 16+ string lengths and use them only for keys longer than 16 :)

File: components/uniset/src/provider.rs (hunk 6 above)
Comment (echeran) [reply components/uniset/src/provider.rs]: Done.

File: components/uniset/src/provider.rs (hunk 7 above)
Comment (sffc) [reply components/uniset/src/provider.rs:322]: I'm sorry, I really don't understand what the problem is.  Is there an *example* of a `<propname>=<propval>` that is *not* unique when truncated to 16 characters?

Option 3 would be to create a very simple hashing function that maps `<propname>=<propval>` that is 16 characters long.

It's essential that ResourceKeys are const-constructible and easy to scan for in assembly code, so revisiting that type is not an option I want to explore until all other options have been exhausted.

Comment (sffc) [reply components/uniset/src/provider.rs:322]: To be clear: when I say "truncate" I mean literally slicing off all characters after the 16th character.  "foo=abcdefghijklmnop" truncates to "foo=abcdefghijkl".

Comment (echeran) [reply components/uniset/src/provider.rs:322]: There's `InPC=Top_And_Bottom_And_Left` & `InPC=Top_And_Bottom_And_Right`; `jg=Manichaean_Sadhe` & `jg=Manichaean_Samekh`; `jg=Manichaean_Ten` & `jg=Manichaean_Teth`. I'm not sure if there are any other instances where truncating is ambiguous.

Comment (sffc) [reply components/uniset/src/provider.rs:322]: Okay. I'm convinced. I tried a few other hashing functions and couldn't reliably eliminate the collisions.  Just make a map (perhaps using Manish's new VecMap) that lives inside provider_ppucd that maps from strings that exceed 16 characters to the fixed strings in the uniset::props provider.

Comment (sffc) [reply components/uniset/src/provider.rs:322]: @Manishearth suggested the PHF crate as another alternative.

Comment (sffc) [reply components/uniset/src/provider.rs:322]: @zbraniecki and @Manishearth also suggested using a codegen'd enum.

File: components/uniset/src/enum_prop_mapping.rs [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+        _ => None,
+    };
+    let name_enum_i32_opt = name_enum_opt.map(|x| x as i32);
+    match (name_enum_i32_opt, val_enum_i32_opt) {
+        (Some(name_i32), Some(val_i32)) => Some((name_i32, val_i32)),
+        _ => None,
+    }
+}
+
+pub fn get_prop_name_identifier(prop_name: &str, prop_val: &str) -> Option<String> {
```
Comment (sffc) [inline components/uniset/src/enum_prop_mapping.rs:824]: This can return a TinyString16

File: components/uniset/src/props.rs [hunk 9]
```diff
@@ -30,373 +31,1772 @@ fn get_prop<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(
     ppucd_property.try_into()
 }
 
-/// Upper
-pub fn upper<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(
+//
```
Comment (sffc) [inline components/uniset/src/props.rs:34]: Optional: move this file to *props/mod.rs* and make a custom *rustfmt.toml* that increases line length so that these match statements don't wrap weirdly.

File: components/uniset/src/enum_props.rs [hunk 10]
```diff
@@ -0,0 +1,673 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum EnumeratedProperty {
```
Comment (sffc) [inline components/uniset/src/enum_props.rs:6]: Are there well-defined discriminant values for these properties?  Take the ones from ICU4C?

File: components/uniset/src/enum_prop_mapping.rs [hunk 11]
```diff
@@ -0,0 +1,876 @@
...
+use litemap::LiteMap;
+
+//
+// Single getter function for enumerated property name:
+// Enum prop name string -> Rust enum
+//
+
+fn get_enum_property_enum(name: &str) -> Option<EnumeratedProperty> {
+    let mut m: LiteMap<&str, EnumeratedProperty> = LiteMap::new();
+    m.insert("bc", EnumeratedProperty::BidiClass);
```
Comment (sffc) [inline components/uniset/src/enum_prop_mapping.rs:15]: Use a rust match on strings

File: components/uniset/src/enum_prop_mapping.rs [hunk 12]
```diff
@@ -0,0 +1,876 @@
...
+    m.insert("WJ", LineBreak::WordJoiner);
+    m.insert("XX", LineBreak::Unknown);
+    m.insert("ZW", LineBreak::ZWSpace);
+    m.insert("ZWJ", LineBreak::ZWJ);
+    m.get(name).cloned()
+}
+
+fn get_lead_canonical_combining_class_enum(name: &str) -> Option<LeadCanonicalCombiningClass> {
+    let mut m: LiteMap<&str, LeadCanonicalCombiningClass> = LiteMap::new();
+    m.insert("0", LeadCanonicalCombiningClass::NotReordered);
```
Comment (sffc) [inline components/uniset/src/enum_prop_mapping.rs:529]: Optional: For these functions, the string input is mapped to an identical integer output. Consider parsing the string to an int, rather than performing the string match. However, you will still need to ensure that the int is valid, since the ints are sparse.

File: components/uniset/src/enum_prop_mapping.rs (hunk 8 above)
Comment (echeran) [reply components/uniset/src/enum_prop_mapping.rs]: Done. At least here, I changed this function to return `Option<TinyStr16>` instead of `Option<String>`.

I thought the long-term intention was that we could, in the future, plumb the `TinyStr16` returned here to be re-used as the identifier string for the `UnicodeProperty` struct returned by the data provider. I at least tried following through that line of thinking, but ran into the problem that if I change the id field of `UnicodeProperty` from `String` to `TinyStr16`, the error is that TinyStr is not currently serde serializable:

```
the trait bound `tinystr::TinyStr16: provider::_::_serde::Serialize` is not satisfied

the trait `provider::_::_serde::Serialize` is not implemented for `tinystr::TinyStr16`

note: required because of the requirements on the impl of `provider::_::_serde::Serialize` for `std::borrow::Cow<'_, tinystr::TinyStr16>`
note: required by `provider::_::_serde::ser::SerializeStruct::serialize_field`
```

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 13]
```diff
@@ -219,7 +294,110 @@ fn get_binary_prop_unisets<'s>(
...
+                        .unwrap()
+                        .get_mut(canonicalized_val_name)
+                        .unwrap();
+                    enum_val_uniset_builder.add_char(std::char::from_u32(*code_point).unwrap());
+                }
+            }
+        }
+    }
+
+    let mut result: HashMap<Cow<'s, str>, UnicodeSet> = HashMap::new();
```
[the remark this answers is not in the document]
Comment (echeran) [reply components/provider_ppucd/src/parse_ppucd.rs]: Done.

File: components/uniset/src/props.rs (hunk 9 above)
Comment (echeran) [reply components/uniset/src/props.rs]: Done.

File: components/uniset/src/enum_props.rs (hunk 10 above)
Comment (echeran) [reply components/uniset/src/enum_props.rs:6]: Filed followup issue #537 for this.

File: components/uniset/src/enum_prop_mapping.rs (hunk 11 above)
Comment (echeran) [reply components/uniset/src/enum_prop_mapping.rs]: Done.

File: components/uniset/src/enum_prop_mapping.rs (hunk 12 above)
Comment (echeran) [reply components/uniset/src/enum_prop_mapping.rs]: Probably don't want to go to this level of optimization because it applies to only some of the enum variants in only the `ccc`, `lccc`, and `tccc` properties. But the ideas of completeness / consistency of enum discriminant values for identifiers for binary props (#536) and matching enum discriminant values used in ICU4C (#537) still make sense.

File: components/provider/tests/data_provider.rs [hunk 14]
```diff
@@ -225,10 +225,7 @@ fn test_warehouse_owned_dyn_generic() {
 fn test_warehouse_owned_dyn_erased_alt() {
     let warehouse = get_warehouse(DATA);
     let response = get_payload_alt(&warehouse as &dyn ErasedDataProvider);
-    assert!(matches!(
-        response,
-        Err(DataError::UnsupportedResourceKey { .. })
-    ));
+    assert!(matches!(response, Err(DataError::UnsupportedResourceKey { .. })));
```
Comment (sffc) [inline components/provider/tests/data_provider.rs:228]: I was getting this diff, too, but running `rustup update` fixes it.

File: components/uniset/src/enum_prop_mapping.rs [hunk 15]
```diff
@@ -0,0 +1,862 @@
+// This file is part of ICU4X. For terms of use, please see the file
```
Comment (sffc) [inline components/uniset/src/enum_prop_mapping.rs:1]: Didn't we decide to move this file into provider_ppucd until we need it outside?

File: components/uniset/src/props/mod.rs [hunk 16]
```diff
@@ -0,0 +1,1802 @@
...
+        CanonicalCombiningClass::Below => {
+            get_prop(provider, key::CANONICAL_COMBINING_CLASS_BELOW_V1)
+        }
+        CanonicalCombiningClass::BelowRight => {
+            get_prop(provider, key::CANONICAL_COMBINING_CLASS_BELOW_RIGHT_V1)
+        }
+        CanonicalCombiningClass::Left => get_prop(provider, key::CANONICAL_COMBINING_CLASS_LEFT_V1),
+        CanonicalCombiningClass::Right => {
+            get_prop(provider, key::CANONICAL_COMBINING_CLASS_RIGHT_V1)
+        }
```
Comment (sffc) [inline components/uniset/src/props/mod.rs:634]: How come there's still inconsistent formatting despite the new line length config file?

File: components/uniset/src/provider.rs [hunk 17]
```diff
@@ -14,78 +14,811 @@ pub mod key {
...
+    pub const CANONICAL_COMBINING_CLASS_CCC31_V1: ResourceKey = resource_key!(uniset, "2=31", 1);
+    pub const CANONICAL_COMBINING_CLASS_CCC32_V1: ResourceKey = resource_key!(uniset, "2=32", 1);
+    pub const CANONICAL_COMBINING_CLASS_CCC33_V1: ResourceKey = resource_key!(uniset, "2=33", 1);
+    pub const CANONICAL_COMBINING_CLASS_CCC34_V1: ResourceKey = resource_key!(uniset, "2=34", 1);
+    pub const CANONICAL_COMBINING_CLASS_CCC35_V1: ResourceKey = resource_key!(uniset, "2=35", 1);
+    pub const CANONICAL_COMBINING_CLASS_CCC36_V1: ResourceKey = resource_key!(uniset, "2=36", 1);
+    pub const CANONICAL_COMBINING_CLASS_HAN_READING_V1: ResourceKey =
+        resource_key!(uniset, "2=6", 1);
+    pub const CANONICAL_COMBINING_CLASS_NUKTA_V1: ResourceKey = resource_key!(uniset, "2=7", 1);
+    pub const CANONICAL_COMBINING_CLASS_KANA_VOICING_V1: ResourceKey =
```
Comment (sffc) [inline components/uniset/src/provider.rs:187]: Optional: make the new rustfmt.toml file apply to this file, too

**Review by sffc — CHANGES_REQUESTED** [review]
No more big things!  Just some small things.

File: components/provider/tests/data_provider.rs (hunk 14 above)
Comment (echeran) [reply components/provider/tests/data_provider.rs]: _Done._

File: components/uniset/src/enum_prop_mapping.rs (hunk 15 above)
Comment (echeran) [reply components/uniset/src/enum_prop_mapping.rs]: Done. Yes, I forgot about that, thanks.

File: components/uniset/src/props/mod.rs (hunk 16 above)
Comment (echeran) [reply components/uniset/src/props/mod.rs]: Done -- I'm not sure why `rustfmt.toml` seemed to get ignored after refactoring `props.rs` -> `props/mod.rs` and adding the .toml config into the new directory. But it does work as expected when I keep the .toml config file in the original directory, which is the parent of that `props` dir (`components/uniset/src`). So I reverted the work to create the `props` dir.

File: components/uniset/src/provider.rs (hunk 17 above)
Comment (echeran) [reply components/uniset/src/provider.rs]: Done. I reverted the change of `props.rs` -> `props/mod.rs` because `rustfmt.toml` wasn't being used by the Rust formatter, and because the Rust formatter config applies not only to `props.rs`, but also here to `provider.rs` and another file in the same directory.

File: components/uniset/src/provider.rs [hunk 18]
```diff
@@ -14,78 +14,633 @@ pub mod key {
...
+    pub const WORD_BREAK_ZWJ_V1: ResourceKey = resource_key!(uniset, "22=22", 1);
 }
 
 #[derive(Debug, PartialEq, Clone)]
-#[cfg_attr(
-    feature = "provider_serde",
-    derive(serde::Serialize, serde::Deserialize)
-)]
+#[cfg_attr(feature = "provider_serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "testing", derive(Hash, Eq))]
```
Comment (sffc) [inline components/uniset/src/provider.rs:643]: Based on feedback from fellow Rustaceans, please just unconditionally implement Hash and Eq, and remove the "testing" feature.

File: components/uniset/src/props.rs [hunk 19]
```diff
@@ -1,26 +1,21 @@
 // This file is part of ICU4X. For terms of use, please see the file
 // called LICENSE at the top level of the ICU4X source tree
-// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
```
Comment (sffc) [inline components/uniset/src/props.rs:3]: Use "main" not "master"

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/uniset/src/provider.rs (hunk 18 above)
Comment (echeran) [reply components/uniset/src/provider.rs]: Done.

**Review by sffc — DISMISSED** [review]
Approved!

According to ICU4X review practices, let's wait for @zbraniecki to leave some additional feedback or remove himself from the reviewers list.

File: components/uniset/src/enum_prop_mapping.rs (hunk 8 above)
Comment (sffc) [reply components/uniset/src/enum_prop_mapping.rs]: You need to enable the "serde" feature on the "tinystr" dependency for this to work.  But let's resolve that later. #550

File: components/provider_ppucd/src/enum_prop_mapping.rs [hunk 20]
```diff
@@ -0,0 +1,862 @@
...
+        "NFD_QC" => Some(EnumeratedProperty::NFDQuickCheck),
+        "NFKC_QC" => Some(EnumeratedProperty::NFKCQuickCheck),
+        "NFKD_QC" => Some(EnumeratedProperty::NFKDQuickCheck),
+        "nt" => Some(EnumeratedProperty::NumericType),
+        "SB" => Some(EnumeratedProperty::SentenceBreak),
+        "tccc" => Some(EnumeratedProperty::TrailCanonicalCombiningClass),
+        "vo" => Some(EnumeratedProperty::VerticalOrientation),
+        "WB" => Some(EnumeratedProperty::WordBreak),
+        _ => None,
+    }
```
Comment (zbraniecki) [inline components/provider_ppucd/src/enum_prop_mapping.rs:40]: `suggestion`: I don't know if you generated it or wrote manually, but it's often more readable to put the `Option` monad outside of the match clause, for example by:

```rust
fn foo(x: &str) -> Option<Kind> {
    Some(
        match x {
            "foo" => Kind::Variant1,
            "bar" => Kind::Variant2,
            _ => return None,
        }
    )
}
```

it may seem small but on files like this it scales and the `Some` by almost all options accumulates both in size and in read cognitive load.

File: components/provider_ppucd/src/parse_ppucd.rs [hunk 21]
```diff
@@ -447,4 +652,472 @@ mod gen_properties_test {
...
+        exp_uni_props_set.insert(hst_na);
+        let gr_link = UnicodeProperty {
+            name: Cow::Borrowed("Gr_Link"),
+            inv_list: vec![5908, 5909],
+        };
+        exp_uni_props_set.insert(gr_link);
+        let bpt_n = UnicodeProperty {
+            name: Cow::Borrowed("1=1"),
+            inv_list: vec![5888, 5901, 5902, 5909],
+        };
```
Comment (zbraniecki) [inline components/provider_ppucd/src/parse_ppucd.rs:1090]: `option`: for maintenance convenience reasons, I often use `From` for scenarios like this to end up with:

```rust
let bpt_n = ("1=1", vec![5888, 5901, 5902, 5909]).into();
```

or, in case of tests even:

```rust
let samples = &[
    ("1=1", vec![5888, 5901, 5902, 5909]),
    ...
];
for elem in samples {
  test(elem.into());
}
```

File: components/uniset/src/.rustfmt.toml [hunk 22]
```diff
@@ -0,0 +1,5 @@
+# This file is part of ICU4X. For terms of use, please see the file
+# called LICENSE at the top level of the ICU4X source tree
+# (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+max_width = 200                    # length of line
```
Comment (zbraniecki) [inline components/uniset/src/.rustfmt.toml:5]: `option`: is there a way to annotate a single file like this? I know that for generated files I can set `cargo clippy` lints, but I'm wondering if there's a better way than `rustfmt.toml` to tell `rust fmt` to back off.

File: components/uniset/src/props.rs [hunk 23]
```diff
@@ -3,24 +3,19 @@
...
 use icu_provider::prelude::*;
 use std::borrow::Cow;
 use std::convert::TryInto;
 
 // helper fn
-fn get_prop<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(
-    ppucd_provider: &D,
-    resc_key: ResourceKey,
-) -> Result<UnicodeSet, UnicodeSetError> {
+fn get_prop<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(ppucd_provider: &D, resc_key: ResourceKey) -> Result<UnicodeSet, UnicodeSetError> {
```
Comment (zbraniecki) [inline components/uniset/src/props.rs:14]: `option`: for maintenance:

```rust
type Result = Result<UnicodeSet, UnicodeSetError>;

fn get_prop<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(ppucd_provider: &D, resc_key: ResourceKey) -> Result {}
```

or even:

```rust
macro_rules! helper_fn {
    ($name:ident, $body:expr) => {
fn $name<'d, D: DataProvider<'d, UnicodeProperty<'d>> + ?Sized>(ppucd_provider: &D, resc_key: ResourceKey) -> Result<UnicodeSet, UnicodeSetError> {
    get_prop(provider, $key)
}
    };
}
```

and then:

```rust
helper_fn!(get_ascii_hex_digit_property, key::ASCII_HEX_DIGIT_V1);
helper_fn!(get_alnum_property, key::ALNUM_V1);
```

**Review by zbraniecki — DISMISSED** [review]
lgtm. Left couple suggestions for making the repetitive code easier to maintain, but no blockers. land at will!

File: components/provider_ppucd/src/enum_prop_mapping.rs (hunk 20 above)
Comment (echeran) [reply components/provider_ppucd/src/enum_prop_mapping.rs]: Done.

File: components/provider_ppucd/src/parse_ppucd.rs (hunk 21 above)
Comment (echeran) [reply components/provider_ppucd/src/parse_ppucd.rs:1090]: I held off on this because I find the syntax for defining structs is more readable -- it includes the field names before the instance value => named arguments => not position-based like a tuple. And therefore more maintainable. Since named arguments are simpler, I'm okay with the the verbosity tradeoff.

File: components/uniset/src/.rustfmt.toml (hunk 22 above)
Comment (echeran) [reply components/uniset/src/.rustfmt.toml:5]: Good question. I checked again, and it looks like rustfmt tries to steer away from such granular configuration because it ["tries to work on as much Rust code as possible"](https://github.com/rust-lang/rustfmt#limitations). In spite of that, I tried applying the attributes mentioned in the [Tips section](https://github.com/rust-lang/rustfmt#tips), but the behavior allowed is very limited and did not come close to what we are thinking of. So currently, the answer is still no.

File: components/uniset/src/props.rs (hunk 23 above)
Comment (echeran) [reply components/uniset/src/props.rs]: Done.

File: components/uniset/src/.rustfmt.toml (hunk 22 above)
Comment (sffc) [reply components/uniset/src/.rustfmt.toml:5]: I saw this in @gregtatum's PR#480: `#[rustfmt::skip]`

I found documentation for it: https://github.com/rust-lang/rustfmt#tips

It looks like you might be able to disable a certain rule for a certain file like this?

```rust
#[rustfmt::skip::macros(html)]
```

And I wonder if changing `#[` to `#![` would make it work for a whole file.

Comment (echeran) [reply components/uniset/src/.rustfmt.toml:5]: I had already tried the `#![` syntax at the file level, and that didn't work. rustfmt only exposes `rustfmt::skip::macros` and `rustfmt::skip::attributes`. Hmm, I see now that there is skip formatting altogether with `rustfmt::skip`. I don't see anything like clippy where rustfmt would allow us to have the granularity of control to turn off a specific type of check (ex: `max_width`).

Comment (sffc) [reply components/uniset/src/.rustfmt.toml:5]: Hmm.  It might be nice if we put whatever code that doesn't want to be formatted in its own files (with as little other code in those files as possible), and then just skipped the whole file.

Could we also annotate every line or every function with `rustfmt::skip`?

Let's discuss tomorrow.

File: components/uniset/src/props.rs (hunk 23 above)
Comment (zbraniecki) [reply components/uniset/src/props.rs]: I see you didn't decide to go for the macro. Fine by me, but more hassle for maintenance later.

**Review by zbraniecki — APPROVED** [review]

File: components/uniset/src/.rustfmt.toml (hunk 22 above)
Comment (sffc) [reply components/uniset/src/.rustfmt.toml:5]: Do we want to still follow up on the rustfmt question?
